### ISSUE TITLE
Predictions-GTs matching in validation

### ISSUE THREAD

**Issue by amichayfeldman** [ultralytics/ultralytics#22585]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

Matches contains sorted prediction–GT index pairs (highest IoU first). The current logic uses np.unique twice — first to remove duplicate GT indices and then duplicate prediction indices. But in some cases this produces incorrect filtering.

Example:

matches = [[0, 0],
           [0, 1],
           [1, 1]]

Expected result:

[[0, 0],
 [1, 1]]

Actual result after the current unique operations:

[[0, 0]]

A greedy matching approach that resolves both prediction and GT conflicts simultaneously would preserve the correct highest-IoU pairs.
https://github.com/ultralytics/ultralytics/blob/38c86c798c9469cf736d9a02c6b788423eb3b5e0/ultralytics/engine/validator.py#L297-L306

Here is my alternative implementation:
```
  if matches.shape[0]:
          keep = np.ones(len(matches), dtype=bool)
          gt_seen = {}
          pred_seen = {}
          for idx, (gt_idx, pred_idx) in enumerate(matches):
              if gt_idx in gt_seen or pred_idx in pred_seen:
                  keep[idx] = False
              else:
                  gt_seen[gt_idx] = True
                  pred_seen[pred_idx] = True
          matches = matches[keep]
      correct[matches[:, 1].astype(int), i] = True
```

### Environment

Ultralytics 8.3.163 🚀 Python-3.11.13 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 10822MiB)
Setup complete ✅ (64 CPUs, 251.3 GB RAM, 604.1/1756.3 GB disk)

OS                  Linux-5.4.0-216-generic-x86_64-with-glibc2.31
Environment         Linux
Python              3.11.13
Install             git
Path                *********
RAM                 251.27 GB
Disk                604.1/1756.3 GB
CPU                 Intel Xeon Gold 6130 2.10GHz
CPU count           64
GPU                 NVIDIA GeForce RTX 2080 Ti, 10822MiB
GPU count           7
CUDA                12.1

setuptools          ✅ 65.5.0
openai-clip         ✅ 1.0.1
lap                 ✅ 0.5.12>=0.5.12
albumentations      ✅ 2.0.8
opencv-python-headless✅ 4.11.0.86
google-cloud-storage✅ 3.1.1>=1.13.2
streamlit           ✅ 1.46.1
clearml             ✅ 2.0.2>=1.16.5
shapely             ✅ 2.1.1>=2.0.6
onnxslim            ✅ 0.1.61>=0.1.43
numpy               ✅ 1.24.4>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.3.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.4>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.4.1>=2.1.1
torch               ✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
onnxruntime-gpu     ✅ 1.22.0>=1.19.2
pytest              ✅ 8.4.1>=8.4.1
faster-coco-eval    ✅ 1.6.7>=1.6.7
openvino            ✅ 2025.2.0>=2025.2.0
flask               ✅ 3.1.1>=3.1.1
coremltools         ✅ 8.3.0>=8.3.0
pynvml              ✅ 12.0.0>=12.0.0
tensorrt            ✅ 10.13.0.35>=10.12.0.36
{'OS': 'Linux-5.4.0-216-generic-x86_64-with-glibc2.31', 'Environment': 'Linux', 'Python': '3.11.13', 'Install': 'git', 'Path': ********, 'RAM': '251.27 GB', 'Disk': '604.1/1756.3 GB', 'CPU': 'Intel Xeon Gold 6130 2.10GHz', 'CPU count': 64, 'GPU': 'NVIDIA GeForce RTX 2080 Ti, 10822MiB', 'GPU count': 7, 'CUDA': '12.1', 'Package Info': {'setuptools': '✅ 65.5.0', 'openai-clip': '✅ 1.0.1', 'lap': '✅ 0.5.12>=0.5.12', 'albumentations': '✅ 2.0.8', 'opencv-python-headless': '✅ 4.11.0.86', 'google-cloud-storage': '✅ 3.1.1>=1.13.2', 'streamlit': '✅ 1.46.1', 'clearml': '✅ 2.0.2>=1.16.5', 'shapely': '✅ 2.1.1>=2.0.6', 'onnxslim': '✅ 0.1.61>=0.1.43', 'numpy': '✅ 1.24.4>=1.23.0', 'matplotlib': '✅ 3.10.3>=3.3.0', 'opencv-python': '✅ 4.11.0.86>=4.6.0', 'pillow': '✅ 11.3.0>=7.1.2', 'pyyaml': '✅ 6.0.2>=5.3.1', 'requests': '✅ 2.32.4>=2.23.0', 'scipy': '✅ 1.15.3>=1.4.1', 'torch': '✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.19.1>=0.9.0', 'tqdm': '✅ 4.67.1>=4.64.0', 'psutil': '✅ 7.0.0', 'py-cpuinfo': '✅ 9.0.0', 'pandas': '✅ 2.2.3>=1.1.4', 'seaborn': '✅ 0.13.2>=0.11.0', 'ultralytics-thop': '✅ 2.0.14>=2.0.0', 'onnxruntime-gpu': '✅ 1.22.0>=1.19.2', 'pytest': '✅ 8.4.1>=8.4.1', 'faster-coco-eval': '✅ 1.6.7>=1.6.7', 'openvino': '✅ 2025.2.0>=2025.2.0', 'flask': '✅ 3.1.1>=3.1.1', 'coremltools': '✅ 8.3.0>=8.3.0', 'pynvml': '✅ 12.0.0>=12.0.0', 'tensorrt': '✅ 10.13.0.35>=10.12.0.36'}}

### Minimal Reproducible Example

matches = [[0, 0],
[0, 1],
[1, 1]]

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @amichayfeldman, thank you for your detailed report and for sharing your idea to improve matching logic 🚀 This is an automated response to help move things forward quickly. An Ultralytics engineer will also review and assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Your toy “matches” illustration is helpful 👍, but to reproduce and validate changes on our side, could you please share:
- A short, self-contained snippet that runs `val` and produces the incorrect filtering (ideally using a tiny dataset or synthetic labels/preds)
- Exact Ultralytics version used (and confirm results after upgrading below)
- Relevant validator args (e.g., IoU/Conf thresholds, NMS settings) and any custom modifications
- A small log excerpt from `val` showing the affected sample(s)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for the clear example and suggested approach 🙏 If you’re open to it, feel free to submit a small PR with unit tests that demonstrate the corrected behavior on your case—this will help us review and validate quickly.

**Comment by Y-T-G** [comment]
Does it improve mAP?

**Comment by amichayfeldman** [comment]
> Does it improve mAP?

It depends on the specific case. It may improve or reduce mAP, but it will certainly change the value, since more bounding boxes/keypoints are included in the computation (increasing recall).

**Comment by glenn-jocher** [comment]
Thanks, agreed—the two-step `np.unique` in validator `match_predictions` can drop a valid pair in cases like this; a single greedy pass over IoU‑sorted matches keeps a proper one‑to‑one mapping. If you can open a PR applying that change in `ultralytics/engine/validator.py` with a tiny test for the `[[0,0],[0,1],[1,1]]` case and a brief before/after mAP note, we’ll review; for context see the [match_predictions reference](https://docs.ultralytics.com/reference/engine/validator/#match_predictions). Note the optional SciPy/Hungarian path (`use_scipy=True`) is known to alter mAP, so it will remain off by default.
